[Wu Xiao Yun] iP

Xiaoyunnn opened the pull request:
# DukeBai

> “In order to seek one’s own direction, one must simplify the mechanics of ordinary, everyday life.” - Plato ([source](https://www.minimalray.com/blog/minimalist-quotes))

DukeBai simplifies the tracking of your upcoming tasks. It's,
- text-based
- easy to learn
- ~~FAST~~ _**SUPER** FAST_ to use



All you need to do is,
1. download it from [here](https://github.com/Xiaoyunnn/ip/releases/tag/A-Jar).
2. double-click the *.jar* file.
3. add your tasks using the command-line interface (CLI).
   - Note: type `help` to see the list of available commands.
4. free your mind and let it manage your tasks for you 🥳

What's more? It is totally **FREE**! 🎉



## Features:
- [x] Managing tasks (todo, deadline, event)
- [ ]  CLI GUI
- [ ] ~~Managing deadlines (coming soon)~~
- [ ] ~~Reminders (coming soon)~~



## Main Method

```java
public class Duke {
    // fields and methods for loading and saving tasks as well as reading user inputs.

    /**
     * Drives the Duke program.
     *
     * @param args Arguments that the user inputs.
     */
    public static void main(String[] args) {
        // creates a folder named "data" in the current directory and 
        // a file "tasks.txt" to save the tasks you have added.
        new Duke("data/tasks.txt").run();
    }
}
```

You may checkout the source code [here](https://github.com/Xiaoyunnn/ip).

Commit:
Add Gradle support

Xiaoyunnn committed:
Add Level-1

Xiaoyunnn committed:
Add Level-2

Xiaoyunnn committed:
Add Level-3

Xiaoyunnn committed:
Add Level 3

Xiaoyunnn committed:
Add Level 5

Xiaoyunnn committed:
Add Level 6 with array ver

Xiaoyunnn committed:
Add Level 6 with arraylist ver

Xiaoyunnn committed:
Add javadoc for DukeException

Xiaoyunnn committed:
Remove Nullpointer exception

Xiaoyunnn committed:
Clean up some code

Xiaoyunnn committed:
Add Level-7

Xiaoyunnn committed:
Merge branch 'branch-Level-7'

Xiaoyunnn committed:
Add Level-8

Xiaoyunnn committed:
Merge branch 'branch-Level-8'
Add Level-8 to change the commands for date using String to
LocalDateTime.

Xiaoyunnn committed:
Add OOP draft

Xiaoyunnn committed:
Add more OOP

Xiaoyunnn committed:
Organize into packages

Xiaoyunnn committed:
Add JUnit Tests

Xiaoyunnn committed:
Add JavaDoc

Xiaoyunnn committed:
Merge branch 'branch-A-JavaDoc'
Add JavaDoc.

Xiaoyunnn committed:
Edit code to comply with coding standards

Xiaoyunnn committed:
Merge branch 'branch-A-CodingStandard'
Edit code to comply with coding stardards.

Xiaoyunnn committed:
Add Level-9

Xiaoyunnn committed:
Merge branch 'branch-Level-9'
Add Level-9.

Xiaoyunnn committed:
Add find command to help list

BryannYeap commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,102 @@
+package duke;
+
+import duke.tasks.Deadline;
+import duke.tasks.Event;
+import duke.tasks.Task;
+import duke.tasks.Todo;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Scanner;
```
I don't think its compulsory but this is the import order used in AB3: 
https://se-education.org/guides/tutorials/intellijCodeStyle.html#tweak-import-order

Damith mentioned in the forum that we could follow this for iP as well. Just a consideration!

BryannYeap commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,28 @@
+ * Exception specifically related to Duke program.
+ */
+public class DukeException extends Exception {
+    private final String message;
```
Should this be in uppercase since it is a constant?

BryannYeap commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,52 @@
+package duke;
+
+import duke.commands.*;
```
Similar to importing java classes, it was mentioned by Damith in the forum that wild card imports should not be used for importing packages as well. If you want to, you can see the forum thread [here](https://github.com/nus-cs2103-AY2122S1/forum/issues/73).

BryannYeap commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    /**
+     * Return the number of task items in the list.
```
Similarly to your other javadocs headers (which are well done! 😺 ), headers should start with words in the form `Returns ...`, `Sends ...`, `Adds ...` (not `Return` or `Returnning` etc.)

BryannYeap commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,108 @@
+                + "|  |_|  \\  /  ____  \\  |  |\n"
+                + "|_______/ /__/    \\__\\ |__|\n";
+        System.out.println(logo);
+        System.out.println("Hello! I'm Bai.\n" +
+                "What can I do for you?");
```
Line breaks should come before the `+` operator 😅. Like this:
```java
        System.out.println("Hello! I'm Bai.\n" 
                + "What can I do for you?");
```

BryannYeap commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,108 @@
+     * Prints the exit message when exit command is called.
+     */
+    public void printExitMessage() {
+        System.out.println("====================================================\n" +
+                "Goodbai. Hope to see you again soon! （ ● ___ ●.）" +
+                "\n====================================================");
```
Cute emoji! 😝

BryannYeap commented on src/main/java/duke/tasks/Todo.java:
```diff
@@ -0,0 +1,45 @@
+    /**
+     * Instantiates a Todo object;
+     *
+     * @param description
+     * @param isDone
```
Did you forget to describe these parameters?

BryannYeap commented on src/main/java/duke/tasks/Todo.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    /**
+     * Instantiates a Todo object;
```
You know you've been coding too much when you end sentences with a `;` instead of a `.` 😳

BryannYeap reviewed (commented):
LGTM! Really clean code! All functions and variables (such as boolean variables) are all named as they should be. Just a few minor mistakes! 
Additionally, I'm not sure how important this is but in the website for [coding standards](https://se-education.org/guides/conventions/java/intermediate.html#comments), the @param descriptions, @return descriptions, etc, in the javadocs start with a capital letter, as seen:
```java
/**
 * Returns lateral location of the specified position.
 * If the position is unset, NaN is returned.
 *
 * @param x  X coordinate of position.
 * @param y Y coordinate of position.
 * @param zone Zone of position.
 * @return Lateral location.
 * @throws IllegalArgumentException  If zone is <= 0.
 */
public double computeLocation(double x, double y, int zone)
    throws IllegalArgumentException {
  ...
}
```

You could take that into consideration! Other than that all good!

Xiaoyunnn committed:
Resolve comments from PR reviews

Xiaoyunnn committed:
Merge branch 'add-gradle-support' into branch-A-Gradle
Add Gradle support.

Xiaoyunnn committed:
Set up gradle

Xiaoyunnn committed:
Merge branch 'branch-A-Gradle'
Set up gradle support and checkstyles.

Xiaoyunnn committed:
Practise JavaFX

Xiaoyunnn committed:
Add JavaFX tutorial iteration 1

Xiaoyunnn committed:
Add JavaFX tutorial iteration 2 draft

Xiaoyunnn committed:
Merge master to branch-Level-10

Xiaoyunnn committed:
Add JavaFX tutorial part 3

Xiaoyunnn committed:
Add JavaFX tutorial part 4

HolmesJJ commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,179 @@
+//        Scene scene = new Scene(helloWorld); // Setting the scene to be our Label
+//
+//        stage.setScene(scene); // Setting the stage to show our screen
+//        stage.show(); // Render the stage.
```
Should these lines of dead code need to be removed? 🤔

HolmesJJ commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,179 @@
+//
+//        stage.setScene(scene); // Setting the stage to show our screen
+//        stage.show(); // Render the stage.
+        Font font = new Font("Osaka", 14);
```
```
private static final int FONT_SIZE = 14;
```
Should this magic number be moved to top by using named constant? 🤔 
And I noticed the same issue in several other places below.

HolmesJJ commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,179 @@
+     * @throws Exception if something goes wrong
+     */
+    @Override
+    public void start(Stage stage) throws Exception {
```
Perhaps need to separate this long method to multiple short methods? 🤔

HolmesJJ commented on src/main/java/duke/commands/TodoCommand.java:
```diff
@@ -0,0 +1,59 @@
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        try {
+            if (userCommand.length() < 5) {
```
Should this magic number be moved to top by using named constant? 🤔

HolmesJJ commented on src/main/java/duke/commands/EventCommand.java:
```diff
@@ -0,0 +1,69 @@
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        try {
+            int atIndex = userCommand.indexOf("/at");
+            if (userCommand.length() < 6 || atIndex <= 6) {
```
Should this magic number be moved to top by using named constant? 🤔
And I noticed the same issue in several other places below.

HolmesJJ commented on src/main/java/duke/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,69 @@
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        try {
+            int byIndex = userCommand.indexOf("/by");
+            if (userCommand.length() < 9 || byIndex <= 9) {
```
Should this magic number be moved to top by using named constant? 🤔
And I noticed the same issue in several other places below.

HolmesJJ reviewed (commented):
LGTM! I like your super clear code and all the code structure is very complete. Only a few code suggestions and magic numbers should be moved to top by using named constant. Well done! 😃

Xiaoyunnn committed:
Improve GUI with JavaFX

Xiaoyunnn committed:
Remove unnecessary code

Xiaoyunnn committed:
Merge branch 'branch-Level-10'
Add nice GUI for Bai.

Xiaoyunnn committed:
Edit code in accordance with checkstyle

Xiaoyunnn committed:
Edit code in accordance with checkstyle

Agentum07 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,103 @@
+     *
+     * @param filePath String of file path of the previously saved task.txt file (if any).
+     */
+    public Duke(String filePath) {
```
Shouldn't there be a `throws IOException` here?

Xiaoyunnn replied:
I used try-catch in the method body so didn't declare `throws IOException` in method header

Agentum07 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,103 @@
+        while (!isExit) {
+            try {
+                String fullCommand = ui.readCommand();
+                Command c = Parser.parse(fullCommand);
```
Maybe come up with a more descriptive name than `c`? `command` would be a good choice.

Xiaoyunnn replied:
thks for pointing it out, will change it!

Agentum07 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,103 @@
+     */
+    String getResponse(String input) {
+        try {
+            Command c = Parser.parse(input);
```
Same here. Maybe a better variable name than `c`

Agentum07 commented on src/main/java/duke/MainWindow.java:
```diff
@@ -0,0 +1,53 @@
+        scrollPane.vvalueProperty().bind(dialogContainer.heightProperty());
+    }
+
+    public void setDuke(Duke d) {
```
I'd suggest having javadocs here. Also, instead of `d` have a descriptive name so that I can better understand the context.

Agentum07 reviewed (commented):
LGTM. Good job!

Xiaoyunnn committed:
Edit code following PR peer review

Xiaoyunnn committed:
Add more JavaDoc

MGRL2201 reviewed (commented):
Very clean, well-written, and documented code. Nice!

Xiaoyunnn committed:
Edit GUI

Xiaoyunnn committed:
Ui: add emoticon to welcome message.
Command: add @return statement for execute method.
Gradle: specify Java version to use and change main class name.

Add emoticon in ui welcome message to prettify the user interface.
Add @return statement for a fuller and clearer documentation.
Change main class name in build.gradle for building .jar file.

Xiaoyunnn committed:
Ui: add new method printInvalidTaskNumError.

IndexOutOfBoundsExceptions are common to both DeleteCommand and
DoneCommand when the user enters an invalid task number to mark as done
or delete.

Let's extract the common method body into a new method in the Ui class
named printInvalidTaskNumError to print the error message when user
enter an invalid task number.

This reduces the duplication of code and abides by the DRY principle.

Xiaoyunnn committed:
Merge pull request #1 from Xiaoyunnn/branch-A-CodeQuality

Refactor code in accordance with the code quality

Xiaoyunnn committed:
Storage class: add assertions in method load.
Command and its subclasses: add class fields to define various error
messages.

The method load in the Storage class has some assumptions made (i.e. each task is split
using | with the necessary descriptions and date and will not reach
the default case).

There are some common error messages in the subclasses for Command,
resulting in code duplications.

Let's add assertions to define the assumptions about the program state
so that the run time can verify them and minimise potential bugs.

Let's add class fields in the parent class Command or relevant subclass
to refactor the code for clearer and more readable code.

Xiaoyunnn committed:
Merge pull request #2 from Xiaoyunnn/branch-A-Assertions

Add assertions and refactor code

Xiaoyunnn committed:
Add continuous integration (CI)

Xiaoyunnn committed:
Add B-ViewSchedules extension

Xiaoyunnn committed:
Add C-Sort extension

Xiaoyunnn committed:
Update help command with schedule and sort instructions

Xiaoyunnn committed:
Merge branch 'branch-C-Sort'
Add C-Sort extension.

Xiaoyunnn committed:
Add better GUI

Xiaoyunnn committed:
Update User Guide

Xiaoyunnn committed:
Edit UG

Xiaoyunnn committed:
Add release link in UG

Xiaoyunnn committed:
Edit CI file

Xiaoyunnn committed:
Add link to section for UG

Xiaoyunnn committed:
Add table of contents for UG

Xiaoyunnn committed:
Edit constant names

Xiaoyunnn committed:
Update UG

Xiaoyunnn committed:
Change welcome message

Xiaoyunnn committed:
Edit welcome message

Xiaoyunnn committed:
Edit jar file name in UG

Xiaoyunnn committed:
Remove unused code

angkoonhwee commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,74 @@
+    /**
+     * Instantiates a Duke instance with default file path "data/tasks".
+     */
+    public Duke() {
```
This method is not used, perhaps can be deleted since it is dead code?

angkoonhwee commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,76 @@
+            return new ExitCommand();
+
+        default:
+            throw new DukeException("I'm sorry, but I don't know what that means! D:");
```
Perhaps can put a `private static final String UNKNOWN_COMMAND_MESSAGE = "I'm sorry, but I don't know what that means! D:"` so there won't be magic strings

angkoonhwee reviewed (commented):
Nice and clean code!!! *^u^*

Xiaoyunnn committed:
Remove unnecessary code

Xiaoyunnn committed:
Edit typo in readme
